pkg/server/plugin/upstreamauthority/certmanager/certmanager.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// loadConfig parses and defaults incoming configure requests
+func (p *Plugin) loadConfig(req *spi.ConfigureRequest) (*Config, error) {
+	config := new(Config)
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to decode configuration file: %s", err)
+	}
+
+	if len(config.IssuerName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "configuration has empty issuer_name property")
+	}
+	// If no issuer_kind given, default to Issuer
+	if len(config.IssuerKind) == 0 {
+		p.log.Warn("configuration has empty issuer_kind property, defaulting to 'Issuer'")
+		config.IssuerKind = "Issuer"
+	}
+	// If no issuer_group given, default to cert-manager.io
+	if len(config.IssuerGroup) == 0 {
+		p.log.Warn("configuration has empty issuer_group property, defaulting to 'cert-manager.io'")
+		config.IssuerGroup = "cert-manager.io"
+	}
+	if len(config.KubeConfigFilePath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "configuration has empty kube_config_path property")
+	}
+
+	return config, nil
+}
+
+func (p *Plugin) MintX509CA(request *upstreamauthorityv0.MintX509CARequest, stream upstreamauthorityv0.UpstreamAuthority_MintX509CAServer) error {
+	ctx := stream.Context()
+
+	// Build PEM encoded CSR
+	csrBuf := new(bytes.Buffer)
+	err := pem.Encode(csrBuf, &pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: request.Csr,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Build CertificateRequest object to be created. Use a generated name to
+	// avoid conflicts.
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "spiffe-ca-",
+			Namespace:    p.config.Namespace,
+		},
+		Spec: cmapi.CertificateRequestSpec{
+			Duration: &metav1.Duration{
+				Duration: time.Second * time.Duration(request.PreferredTtl),
+			},
+			Request: csrBuf.Bytes(),
+			Usages: []cmapi.KeyUsage{
+				cmapi.UsageDigitalSignature,
+				cmapi.UsageKeyEncipherment,
+				cmapi.UsageCertSign,
+			},
+			IsCA: true,
+			IssuerRef: cmmeta.ObjectReference{
+				Name:  p.config.IssuerName,
+				Kind:  p.config.IssuerKind,
+				Group: p.config.IssuerGroup,
+			},
+		},
+	}
+	if err := p.cmclient.Create(ctx, cr); err != nil {
+		return err
+	}
+
+	// Poll the CertificateRequest until it is signed. If not signed after 300
+	// polls, error.
+	obj := client.ObjectKey{Name: cr.Name, Namespace: cr.Namespace}
+	for i := 0; true; i++ {
+		if i == 60*5 { // ~5 mins
+			return fmt.Errorf("failed to wait for CertificateRequest %s/%s to become ready: %#+v", cr.Namespace, cr.Name, cr.Status)
+		}
+
+		time.Sleep(time.Second)
+
+		if err := p.cmclient.Get(ctx, obj, cr); err != nil {
+			return err
+		}
+
+		if cmutil.CertificateRequestIsDenied(cr) {
+			cond := cmutil.GetCertificateRequestCondition(cr, cmapi.CertificateRequestConditionDenied)
+			return fmt.Errorf("created CertificateRequest %s/%s has been denied: %s:%s", cr.Namespace, cr.Name, cond.Reason, cond.Message)
```

Comment:
We should probably log this error with the appropriate fields, and then return a more succinct error message to core. That way we'll get structured logging since core won't extract all of this stuff out from the error it receives.

Similar comment for all the `fmt.Errorf` messages that are templating details into the error message